core/drivers/imx_sc_api.c

```diff
@@ -0,0 +1,435 @@
+		return res;
+	}
+
+	RPC_U16(&msg, 0) = (uint16_t)resource;
+	RPC_U8(&msg, 2) = secure_partition;
+
+	res = sc_call_rpc(&msg, true);
+	if (res != TEE_SUCCESS) {
+		EMSG("Communication error");
+		return res;
+	}
+
+	err = msg.header.func;
+	if (err != SC_ERR_NONE) {
+		EMSG("Unable to assign resource, sc_error: %" PRId32,
+		     err);
+		return TEE_ERROR_GENERIC;
+	}
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result imx_sc_rm_enable_jr(unsigned int jr_index)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	enum sc_resource jr_res = SC_RES_LAST;
+	enum sc_resource jr_out_res = SC_RES_LAST;
+
+	switch (jr_index) {
+	case 0:
+		jr_res = SC_RES_CAAM_JR0;
+		jr_out_res = SC_RES_CAAM_JR0_OUT;
+		break;
+
+	case 1:
+		jr_res = SC_RES_CAAM_JR1;
+		jr_out_res = SC_RES_CAAM_JR1_OUT;
+		break;
+
+	case 2:
+		jr_res = SC_RES_CAAM_JR2;
+		jr_out_res = SC_RES_CAAM_JR2_OUT;
+		break;
+
+	case 3:
+		jr_res = SC_RES_CAAM_JR3;
+		jr_out_res = SC_RES_CAAM_JR3_OUT;
+		break;
+
+	default:
+		EMSG("Wrong JR Index, should be 0, 1, 2 or 3");
+		return TEE_ERROR_GENERIC;
+	}
+
+	/* Assign JR resources to secure world */
+	res = sc_rm_assign_resource(jr_res);
+	if (res != TEE_SUCCESS) {
+		EMSG("Assign SC_R_CAAM_JR%" PRId32 " resource failed",
+		     jr_index);
+		return res;
+	}
+
+	res = sc_rm_assign_resource(jr_out_res);
+	if (res != TEE_SUCCESS) {
+		EMSG("Assign SC_R_CAAM_JR%" PRId32 "_OUT resource failed",
+		     jr_index);
+		return res;
+	}
+
+	/* Power ON JR resources */
+	res = sc_pm_set_resource_power_mode(jr_res, SC_PM_PW_MODE_ON);
+	if (res != TEE_SUCCESS) {
+		EMSG("POWER ON SC_R_CAAM_JR%" PRId32 " resource failed",
+		     jr_index);
+		return res;
+	}
+
+	res = sc_pm_set_resource_power_mode(jr_out_res, SC_PM_PW_MODE_ON);
+	if (res != TEE_SUCCESS) {
+		EMSG("POWER ON SC_R_CAAM_JR%" PRId32 "_OUT resource failed",
+		     jr_index);
+		return res;
+	}
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result imx_sc_seco_start_rng(void)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	enum sc_error err = SC_ERR_LAST;
+	enum sc_seco_rng_status status = SC_SECO_RNG_STAT_UNAVAILABLE;
+	unsigned int retry = 0;
+	struct sc_rpc_msg msg = {
+		.header.version = SC_RPC_VERSION,
+		.header.size = 1,
+		.header.svc = SC_RPC_SVC_SECO,
+		.header.func = SC_SECO_FUNC_START_RNG,
+	};
+
+	for (retry = RNG_INIT_RETRY; retry; retry--) {
+		res = sc_call_rpc(&msg, true);
+		if (res != TEE_SUCCESS) {
+			EMSG("Configuration error");
+			return res;
+		}
+
+		err = msg.header.func;
+		if (err != SC_ERR_NONE) {
+			EMSG("RNG status: %" PRId32, err);
+			return TEE_ERROR_GENERIC;
+		}
+
+		status = RPC_U32(&msg, 0);
+
+		if (status == SC_SECO_RNG_STAT_READY)
+			return TEE_SUCCESS;
+	}
+
+	return TEE_ERROR_GENERIC;
+}
+
+register_phys_mem(MEM_AREA_IO_SEC, SC_IPC_BASE_SECURE, SC_IPC_SIZE);
+TEE_Result imx_sc_driver_init(void)
+{
+	vaddr_t va = 0;
+
+	va = core_mmu_get_va(SC_IPC_BASE_SECURE, MEM_AREA_IO_SEC, SC_IPC_SIZE);
+	if (!va)
+		return TEE_ERROR_GENERIC;
+
+	mutex_lock(&scu_mu_mutex);
```

Comment:
are these mutex useful?

Reply:
These mutex have been added in case the SC driver gets re-initialized during ```sc_ipc_read/write()```
This reminds me that mutex should include ```va = core_mmu_get_va() ...``` and ```secure_ipc_addr = va;```